File: AICON Type1 Source Code Sample/angular-chatbot/src/app/core/interceptor/request.interceptor.ts
import { HttpHeaders, HttpInterceptorFn, HttpParams, HttpClient, HttpRequest, HttpHandlerFn } from '@angular/common/http';
import { inject } from '@angular/core';
import { AuthService } from '../../services/auth.service';
import { Observable, switchMap } from 'rxjs';
import { AuthenticationResponse } from '../../dto/authentication-response';
import { Router } from '@angular/router';
import { API_ENDPOINT } from '../api-endpoint';
import { AppConfigService } from '../app-config/app-config.service';

/**
 * Define request interceptor for all http requests
 * @param req : HttpRequest<unknown>
 * @param next : HttpHandlerFn
 * @returns : Observable<HttpEvent<unknown>>
 */
export const requestInterceptor: HttpInterceptorFn = (req, next) => {
  const accessToken = localStorage.getItem('access_token') ?? '';

  if (req.url.includes(API_ENDPOINT.AUTH) || req.url.includes('.json')) {
    // In case the request is for authentication or get configuration file
    return handleRequest(req, next, true);
  } else if (tokenExpired(accessToken)) {
    // If the access token is expired, get a new one using the refresh token

    // Inject dependencies
    const authService = inject(AuthService);

    // Call the function to get a new access token using the refresh token
    return getRefreshToken().pipe(
      switchMap(response => {
        // If the response is successful, save the new access token and refresh token
        authService.saveToken(response.access_token, response.refresh_token);

        // Continue with the request using the new access token
        return handleRequest(req, next, false);
      })
    );
  } else {
    // If the access token is not expired, proceed with the request
    return handleRequest(req, next, false);
  }
};

/**
 * This function handles the request by adding the access token to the headers if it is not for authentication or get configuration file
 * @param req : HttpRequest<unknown>
 * @param next : HttpHandlerFn
 * @param isNeededToken : boolean
 * @returns : Observable<HttpEvent<unknown>>
 */
function handleRequest(req: HttpRequest<unknown>, next: HttpHandlerFn, isNeededToken: boolean) {
  // If the request is for authentication, do not add the access token to the headers

  if (!isNeededToken) {
    // Clone the request and add the access token to the headers
    req = req.clone({
      setHeaders: {
        Authorization: `Bearer ${localStorage.getItem('access_token')}`
      }
    });
  }

  return next(req);
}

/**
 * This function checks if the token is expired or not
 * @param token : string
 * @returns : boolean
 */
function tokenExpired(token: string): boolean {
  if (!token) {
    return true;
  }

  // Split the token into its parts (header, payload, signature)
  const tokenParts = token.split('.');
  if (tokenParts.length !== 3) {
    return true;
  }

  // Decode the payload part of the token (base64url decode)
  const payload = JSON.parse(atob(tokenParts[1]));

  // Convert to milliseconds
  const expirationTime = payload.exp * 1000;

  // Check if the current time is greater than the expiration time
  return Date.now() > expirationTime;
}

/**
 * This function gets a new access token using the refresh token
 * @returns : Observable<AuthenticationResponse>
 */
function getRefreshToken(): Observable<AuthenticationResponse> {
  const refreshToken = localStorage.getItem('refresh_token') ?? '';

  if (tokenExpired(refreshToken)) {
    // If the refresh token is expired, redirect to the login page

    // Inject dependencies
    const authService = inject(AuthService);
    const router = inject(Router);

    // Remove the access token and refresh token from local storage
    authService.removeToken();

    // Redirect to the login page
    router.navigate(['/login']);

    // Return an empty observable to avoid further processing
    return new Observable<AuthenticationResponse>();
  } else {
    // If the refresh token is not expired, proceed to get a new access token

    // Inject dependencies
    const httpClient = inject(HttpClient);
    const appConfigService = inject(AppConfigService);

    // Define the URL for the refresh token endpoint
    const url = appConfigService.get('aiConApiEndpoint') + API_ENDPOINT.AUTH;

    const headers: HttpHeaders = new HttpHeaders({
      'Content-Type': 'application/x-www-form-urlencoded'
    });

    // Set the parameters for the request
    let params = new HttpParams();
    params = params.set('client_id', appConfigService.get('clientID'));
    params = params.set('client_secret', appConfigService.get('clientSecret'));
    params = params.set('grant_type', appConfigService.get('refreshGrantType'));
    params = params.set('refresh_token', localStorage.getItem('refresh_token') ?? '');

    return httpClient.post<AuthenticationResponse>(url, params.toString(), { headers: headers });
  }
}

File: AICON Type1 Source Code Sample/blazor-chatbot/Common/RequestInterceptor.cs
using BlazorChatApp.Service.ChatBot;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using System.Text.Json;

namespace BlazorChatApp.Common
{
    public class RequestInterceptor
    {
        private readonly IJSRuntime JS;
        private readonly IChatBotService ChatBotService;
        private readonly IConfiguration Configuration;
        private readonly NavigationManager NavigationManager;

        // constructor
        public RequestInterceptor(IJSRuntime js,
        IChatBotService chatBotService,
        IConfiguration configuration,
        NavigationManager navigationManager)
        {
            JS = js;
            ChatBotService = chatBotService;
            Configuration = configuration;
            NavigationManager = navigationManager;
        }

        // This function gets a new access token using the get token or refresh token
        public async Task<bool> GetTokenAsync(string? code, bool isRefresh = false)
        {
            try
            {
                // Get clientId and clientSecret from configuration
                var clientId = Configuration["DomainSettings:clientID"];
                var clientSecret = Configuration["DomainSettings:clientSecret"];

                // If it is not isrefresh and code on parameter null, you will return to the login page
                if (!isRefresh && string.IsNullOrEmpty(code))
                {
                    NavigationManager.NavigateTo("/login");
                }
                //Call Api to get token or refresh token using clientId, clientSecret, code and isRefresh
                var responseApi = await ChatBotService.GetTokenAsync(clientId, clientSecret, code, isRefresh);

                if (responseApi == null)
                {
                    return false;
                }

                // Remove old tokens and save new tokens in local storage
                await RemoveTokenAsync();
                // Save the access new token and refresh token in local storage
                await SaveTokenAsync(responseApi.AccessToken, responseApi.RefreshToken);

                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
                return false;
            }
        }
        // Check token in local storage and refresh it
        public async Task<bool> CheckTokenLocalStorage()
        {
            // Get access token from local storage
            var accessToken = await JS.InvokeAsync<string>("localStorage.getItem", "access_token");

            // If access token is null or expired, try to refresh it
            if (TokenExpired(accessToken))
            {
                // Get refresh token from local storage
                var refreshToken = await JS.InvokeAsync<string>("localStorage.getItem", "refresh_token");
                if (!TokenExpired(refreshToken))
                {
                    // If refresh token is valid, get a new access token using the refresh token
                    await GetTokenAsync(refreshToken, true);
                    return true;
                }
                else
                {
                    // If both access token and refresh token are invalid, remove tokens in local storage and redirect to login
                    await RemoveTokenAsync();
                    NavigationManager.NavigateTo("/login");
                    return false;

                }
            }
            return true;
        }

        /// <summary>
        /// Check token expiry date
        /// </summary>
        /// <param name="token"></param>
        /// <returns></returns>
        public bool TokenExpired(string token)
        {
            // If token is null or empty, consider it expired
            if (string.IsNullOrEmpty(token))
            {
                return true;
            }
            // Split the token into parts (header, payload, signature)
            var tokenParts = token.Split('.');
            // If the token does not have exactly 3 parts, consider it invalid
            if (tokenParts.Length != 3)
            {
                return true;
            }

            try
            {
                var payload = tokenParts[1];
                // Base64Url decode
                var paddedPayload = PadBase64(payload);
                var jsonBytes = Convert.FromBase64String(paddedPayload);
                var json = Encoding.UTF8.GetString(jsonBytes);

                using var doc = JsonDocument.Parse(json);
                if (!doc.RootElement.TryGetProperty("exp", out var expElement))
                {
                    return true;
                }

                var exp = expElement.GetInt64(); // exp in seconds
                var expirationTime = DateTimeOffset.FromUnixTimeSeconds(exp).UtcDateTime;
                // Check if the current time is greater than the expiration time
                return DateTime.UtcNow > expirationTime;
            }
            catch
            {
                return true; // If anything fails, assume token is invalid/expired
            }
        }

        /// <summary>
        /// // Decode the payload part of the token (base64url decode)
        /// </summary>
        /// <param name="base64"></param>
        /// <returns></returns>
        private string PadBase64(string base64)
        {
            // Add padding if needed (base64url to base64)
            int padding = 4 - (base64.Length % 4);
            if (padding < 4)
            {
                base64 += new string('=', padding);
            }

            // Replace base64url characters
            return base64.Replace('-', '+').Replace('_', '/');
        }

        /// <summary>
        /// Save the access token and refresh token in local storage
        /// </summary>
        /// <param name="accessToken"></param>
        /// <param name="refreshToken"></param>
        /// <returns></returns>
        public async Task SaveTokenAsync(string accessToken, string refreshToken)
        {
            await JS.InvokeVoidAsync("localStorage.setItem", "access_token", $"{accessToken}");
            await JS.InvokeVoidAsync("localStorage.setItem", "refresh_token", $"{refreshToken}");
        }

        /// <summary>
        /// Remove the access token and refresh token from local storage
        /// </summary>
        /// <returns></returns>
        public async Task RemoveTokenAsync()
        {
            await JS.InvokeVoidAsync("localStorage.removeItem", "access_token");
            await JS.InvokeVoidAsync("localStorage.removeItem", "refresh_token");
        }

        /// <summary>
        /// Get user id from token
        /// </summary>
        /// <param name="token"></param>
        /// <returns></returns>
        public async Task<string?> GetUserIdFromToken(string token)
        {
            try
            {
                var handler = new JwtSecurityTokenHandler();
                var jwtToken = handler.ReadJwtToken(token);
                return jwtToken.Claims.FirstOrDefault(c => c.Type == ClaimTypes.UserData)?.Value;
            }
            catch
            {
                return null;
            }
        }
    }
}
